### DISCUSSION TITLE
Support for object encoding in URL search parameters

### DISCUSSION THREAD

**Discussion by mrtcntn** [vercel/next.js#21180] [Ideas]
# Feature request

## I would like to encode JSON objects into URL's search parameters

I was developing an endpoint to get items filtered and paginated from a collection in a database. The collection contains items type of **Invitation**, which have attributes like "_createdBy_", "_createdAt_", "_invitedEmailAddress_" and "_type_". I wanted to encode following parameters inside the query string:

page = 0
rowsPerPage = 10
filter = {"createdBy": 1, "type": 2} (where 1 is a user id and type is an enum)

To encode filter object, separated its keys and encoded them as following:

filter[createdBy] = 1
filter[type] = 2

Resulting URL is the following, since encoding converts square brackets into %5b and %5d:

http://localhost:3000/api/invitation?page=0&rowsPerPage=10&filter%5bcreatedBy%5d=1&filter%5btype%5d=2

However, this results in the following query object in my API handler's request part:

{"page": 0, "rowsPerPage": 10, "filter[createdBy]": 1, "filter[type]": 2}

Seems that I can't embed objects in the URL search params.

### Desired behavior

For these input

page = 0
rowsPerPage = 10
filter = {"createdBy": 1, "type": 2}

The following output in the query part is desired.
{"page": 0, "rowsPerPage": 10, "filter": {"createdBy": 1, "type": 2}}

Thanks in advance.

**Comment by sm-adil** [comment]
I second this feature request, it would be really helpful if we can embed objects in the URL search params and retrieve the same object in the API handler.

I'm currently trying to migrate our plain old express endpoints to next.js and have bumped into this roadblock. 

@mrtcntn do you have any workaround or suggestion on how you plan to implement similar behavior?

**Reply by mrtcntn** [reply]
I have investigated the difference between Express.js and NextJS and found out that Express.js uses [qs package](https://www.npmjs.com/package/qs). Therefore as a workaround, I also used this package as a wrapping function to my API handlers.

```
import qs from "qs";

export const parseQuery = (fn) => async (req, res) => {
    const questionMarkIdx = req.url.indexOf('?');
    if (questionMarkIdx > 0)
        req.query = {...req.query, ...qs.parse(req.url.substring(questionMarkIdx), {ignoreQueryPrefix: true})};
    return await fn(req, res);
}
```

Doesn't seem like an elegant solution since I don't remove the redundant parts from embedded query object. It is not safe to completely replace the embedded query object since it might contain some parameters from the url (e.g. you have an "id" part on the embedded query object if your API route is something like /post/[id].js) therefore I chose to append qs' parse result to Next's own result. I didn't dig deep into NextJS internals but I think this can be also implemented inside NextJS.
